File: kotlin/lib/src/main/kotlin/desertroad/solar/internal/Angle.kt
package desertroad.solar.internal


@JvmInline
internal value class Angle(private val radianValue: Double) {

    inline val inDegrees: Double get() = Math.toDegrees(radianValue)

    inline val inRadians: Double get() = radianValue

    fun isNaN() = radianValue.isNaN()

    operator fun unaryMinus() = Angle(-radianValue)

    operator fun plus(other: Angle) = Angle(radianValue + other.radianValue)

    operator fun minus(other: Angle) = Angle(radianValue - other.radianValue)

    operator fun times(other: Number) = Angle(radianValue * other.toDouble())

    operator fun div(other: Angle): Double = radianValue / other.radianValue

    companion object {

        val ONE_ROTATION = 360.degrees

        inline val Number.degrees: Angle get() = Angle(Math.toRadians(toDouble()))

        inline val Number.radians: Angle get() = Angle(toDouble())

        inline val Number.rotations: Angle get() = toDouble() * ONE_ROTATION

        operator fun Number.times(angle: Angle) = Angle(toDouble() * angle.radianValue)

        fun sin(angle: Angle) = kotlin.math.sin(angle.radianValue)

        fun cos(angle: Angle) = kotlin.math.cos(angle.radianValue)
    }
}

File: dart/lib/src/angle.dart
import 'dart:math' as dart_math;

class Angle {
  final double _radianValue;

  double get inRadians => _radianValue;

  double get inDegrees => 180.0 / dart_math.pi * _radianValue;

  Angle(this._radianValue);

  Angle operator +(Angle other) => Angle(_radianValue + other._radianValue);

  Angle operator -(Angle other) => Angle(_radianValue - other._radianValue);

  Angle operator *(num other) => Angle(_radianValue * other.toDouble());

  double operator /(Angle other) => _radianValue / other._radianValue;
}

final _oneRotation = 360.degrees;

extension AngleExt on num {
  Angle get degrees => Angle(dart_math.pi / 180.0 * toDouble());

  Angle get radians => Angle(toDouble());

  Angle get rotations => _oneRotation * toDouble();

  Angle operator *(Angle other) => other * toDouble();
}

double sin(Angle angle) => dart_math.sin(angle._radianValue);

double cos(Angle angle) => dart_math.cos(angle._radianValue);
